NuRadioMC/SignalProp/analyticraytracing.py

```diff
@@ -1833,11 +1855,385 @@ def get_path(self, iS, n_points=1000):
+        The trace propagation is explained in this paper: https://link.springer.com/article/10.1140/epjc/s10052-023-11238-y
+
+        Parameters
+        ----------
+        pulse: np.ndarray
+            3d array with the frequency spectrum of np.array([eR, eTheta, ePhi]), usually provided by the apply_propagation_effects function
+        samp_rate: float
+            sampling rate of the time traces
+        i_solution: int
+            choose which ray-tracing solution should be propagated
+        bire_model: string
+            choose the interpolation to fit the measured refractive index data
+            options include (A, B, C, D, E) description can be found under: NuRadioMC/NuRadioMC/utilities/birefringence_models/model_description
+
+        Returns
+        -------
+
+        final pulse: numpy.array([eR, eTheta, ePhi])
+            [0] - eR        - final frequency spectrum of the radial component - not altered by the function
+            [1] - eTheta    - final frequency spectrum of the theta component
+            [2] - ePhi      - final frequency spectrum of the phi component
+        """
+
+        t_fast = base_trace.BaseTrace()
+
+        ice_n = self.__medium
+        ice_birefringence = medium.get_ice_model('birefringence_medium')
+        ice_birefringence.__init__(bire_model)
+
+        acc = int(self.get_path_length(i_solution))
+        path = self.get_path(i_solution, n_points=acc)
+
+        for i in range(acc - 1):
+
+            refractive_index = ice_n.get_index_of_refraction(path[i])
+            refractive_index_birefringence = ice_birefringence.get_birefringence_index_of_refraction(path[i])
+
+            nx, ny, nz = refractive_index + refractive_index_birefringence - 1.78
+            dD = path[i + 1] - path[i]     
+
+            direction = np.array(dD)
+            len_diff = np.linalg.norm(direction)
+            direction = direction / len_diff
+
+            N_effective = self.get_effective_index_birefringence(direction, nx, ny, nz)
+            sky_polarization = self.get_polarization_birefringence(N_effective[0], N_effective[1], direction, nx, ny, nz)
+
+            t_0, t_1 = len_diff * N_effective / (speed_of_light * units.m / units.ns)
+
+            a, b = sky_polarization[0, 1:]
+            c, d = sky_polarization[1, 1:]
+
+            if np.isclose(a * d - b * c, 0) or np.isnan([a, b, c, d]).any():
+                self.__logger.warning("warning: Polarization vectors similar, R-matrix not invertible, iteration" + str(i))
+                continue
+
+            R = np.matrix([[a, b], [c, d]])
+
+            birefringent_base = R * pulse[1:] 
+
+            t_fast.set_frequency_spectrum(birefringent_base[1], sampling_rate=samp_rate)
+            t_fast.apply_time_shift(t_1 - t_0)
+            birefringent_base[1] = t_fast.get_frequency_spectrum()
+            
+            Rinv = np.linalg.inv(R)
```

Comment:
I think this can be sped up by realizing that the matrix is orthogonal and, therefore, the inverse is equal to the transpose of the matrix. I don't know how np.linalg.inv is implemented but it likely uses a numerical approach which should be much slower than a simple transpose operation.

Reply:
Done, updated special cases.